proof_of_stake/src/rewards.rs

```diff
@@ -0,0 +1,85 @@
+//! PoS rewards
+
+use rust_decimal::Decimal;
+use rust_decimal_macros::dec;
+use thiserror::Error;
+
+/// Errors during rewards calculation
+#[derive(Debug, Error)]
+pub enum RewardsError {
+    /// number of votes is less than the threshold of 2/3
+    #[error(
+        "Insufficient votes, needed at least 2/3 of the total bonded stake"
+    )]
+    InsufficentVotes,
+    /// rewards coefficients are not set
+    #[error("Rewards coefficients are not properly set.")]
+    CoeffsNotSet,
+}
+
+/// Holds coefficients for the three different ways to get PoS rewards
+#[derive(Debug, Copy, Clone)]
+#[allow(missing_docs)]
+pub struct PosRewards {
+    pub proposer_coeff: Decimal,
+    pub signer_coeff: Decimal,
+    pub active_val_coeff: Decimal,
+}
+
+/// Holds relevant PoS parameters and is used to calculate the coefficients for
+/// the rewards
+#[derive(Debug, Copy, Clone)]
+pub struct PosRewardsCalculator {
+    proposer_param: Decimal,
+    signer_param: Decimal,
+    signing_stake: u64,
+    total_stake: u64,
+}
+
+impl PosRewardsCalculator {
+    /// Instantiate a new PosRewardsCalculator
+    pub fn new(
+        proposer_param: Decimal,
+        signer_param: Decimal,
+        signing_stake: u64,
+        total_stake: u64,
+    ) -> Self {
+        Self {
+            proposer_param,
+            signer_param,
+            signing_stake,
+            total_stake,
+        }
+    }
+
+    /// Calculate the reward coefficients
+    pub fn get_reward_coeffs(&self) -> Result<PosRewards, RewardsError> {
+        // TODO: think about possibility of u64 overflow
+        let votes_needed = self.get_min_required_votes();
+        if self.signing_stake < votes_needed {
+            return Err(RewardsError::InsufficentVotes);
+        }
+
+        // Logic for determining the coefficients
+        // TODO: error handling to ensure proposer_coeff is > 0?
+        let proposer_coeff = self.proposer_param
+            * Decimal::from(self.signing_stake - votes_needed)
+            / Decimal::from(self.total_stake)
+            + dec!(0.01);
```

Comment:
It's a hard-coded constant to ensure that there is always at least a 1% block reward to the proposer (https://specs.namada.net/economics/proof-of-stake/reward-distribution.html#distribution-of-block-rewards-to-validators).

I could make this an actual parameter somewhere rather than hard-coding.

Reply:
done in 43bf2a7